File: apps/mobile/modules/novella-rs/ios/NovellaRsModule.swift
import ExpoModulesCore
import Foundation

private final class RsOperationFailedException: Exception, @unchecked Sendable {
  let operation: String

  init(_ operation: String) {
    self.operation = operation
    super.init()
  }

  override var reason: String {
    "Novella Rust operation '\(operation)' failed"
  }
}

public final class NovellaRsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NovellaRs")

    AsyncFunction("convertWoff2ToTtf") { (data: Data) throws -> Data in
      guard let result = NovellaRsBridge.convertWoff2ToTtf(data) else {
        throw RsOperationFailedException("convertWoff2ToTtf")
      }
      return result
    }

    AsyncFunction("extractInvisibleCodepoints") { (data: Data) -> [Int] in
      NovellaRsBridge.extractInvisibleCodepoints(data).map { $0.intValue }
    }

  }
}

File: apps/mobile/modules/novella-rs/android/src/main/java/sh/celia/novella/modules/novellars/NovellaRsModule.kt
package sh.celia.novella.modules.novellars

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private class RsOperationFailedException(operation: String) :
  CodedException("Novella Rust operation '$operation' failed")

class NovellaRsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NovellaRs")

    AsyncFunction("convertWoff2ToTtf") { data: ByteArray ->
      NovellaRsBridge.convertWoff2ToTtf(data).also {
        if (it.isEmpty()) throw RsOperationFailedException("convertWoff2ToTtf")
      }
    }

    AsyncFunction("extractInvisibleCodepoints") { data: ByteArray ->
      NovellaRsBridge.extractInvisibleCodepoints(data).toList()
    }

  }
}
